### ISSUE TITLE
dynkomi_interval

### ISSUE THREAD

**Issue by lemonsqueeze** [pasky/pachi#52]
Hi,

Came across this game yesterday: [(sgf)](https://github.com/pasky/pachi/files/383322/KBAC_pachipachi_sgf.txt)

![screenshot](https://cloud.githubusercontent.com/assets/2463904/17130687/221fdfe8-531a-11e6-9a64-8796c1669ac9.png)

Pachi (black) has a sure 2.5pt win, winrate is consistently above 70% (95% for past 2 moves) 
yet inexplicably drops after oponent plays dame, and resigns:  [(full log here)](https://github.com/pasky/pachi/files/383373/pachi.log.gz)

```
IN: genmove b
Move:  261  Komi: 0.5  Handicap: 2  Captures B: 5 W: 11
      A B C D E F G H J K L M N O P Q R S T        A B C D E F G H J K L M N O P Q R S T  
    +---------------------------------------+    +---------------------------------------+
 19 | . X X O . O O X . . . . . . . . . . . | 19 | X X X O O O O X X X X X X X X X X X X |
 18 | . X O O O O X X . . . . X O X . . . . | 18 | X X O O O O X X X X X X X X X X X X X |
 17 | . X X O X O O X X . O . X O X . . X . | 17 | X X X O X O O X X X X X X X X X X X X |
 16 | . X X X X O O O X . X X X X O X X . . | 16 | X X X X X O O O X X X X X X O X X X X |
 15 | X X X O O O . O X . O X O O O O X . . | 15 | X X X O O O O O X X X X O O O O X X X |
 14 | O O X X O)O O X X X X O . . . X O X . | 14 | O O X X , O O X X X X O O O O O O X X |
 13 | O . O X X O X O X O X O . O O O O X . | 13 | O O O X X O X O X O X O O O O O O X X |
 12 | . . O X . X X O O O O O X X X O X X . | 12 | O O O X X X X O O O O O O O O O X X X |
 11 | . . O X X . X O . O X . . . O . O X . | 11 | O O O X X X X O O O O O O O O O O X X |
 10 | O O . O X X O . O O . . . . O . O O X | 10 | O O O O X X O O O O O O O O O O O O X |
  9 | O O . O X O O O O X O . . . O . . O X |  9 | O O O O X O O O O X O O O O O O O O X |
  8 | O X O O O O X O X X O . . . O . O X . |  8 | O X O O O O X O X X O O O O O O O X X |
  7 | X X O X O O X X X O . . . . O O O X . |  7 | X X O X O O X X X O O O O O O O O X X |
  6 | X X X X X X X X X O . . . . O X X X . |  6 | X X X X X X X X X O O O O O O X X X X |
  5 | X O O O X . X O O O . . . . . O O X . |  5 | X O O O X X X O O O O O O O O O O X X |
  4 | O O O X . . X O X O . O O . . O X . . |  4 | O O O X X X X O X O O O O O O O X X X |
  3 | . . O X X O X X X O O O X O O O X . . |  3 | O O O X X X X X X O O O X O O O X X X |
  2 | . O O O X X . . X X O X X X O X X . . |  2 | O O O O X X X X X X O X X X O X X X X |
  1 | . . O O O X . . . . X . X . X . X . . |  1 | O O O O O X X X X X X X X X X X X X X |
    +---------------------------------------+    +---------------------------------------+

prior_dcnn(pass) = [ Q1 K1 A2 B13 O1 ]      [ 0.71 0.10 0.03 0.02 0.01 ]
(UCT tree; root white; extra komi 0.000000; max depth 7)
[pass] 0.232/5004 [prior 0.000/0 amaf 0.000/0 crit -0.768 vloss 0] h=0 c#=115 <26f04c36014ea48e>
 [Q1] 0.096/1461 [prior 0.991/1131 amaf 0.260/6185 crit 0.001 vloss 0] h=0 c#=112 <26f04c42014ea49a>
  [A2] 0.076/1262 [prior 0.278/36 amaf 0.085/4074 crit -0.000 vloss 0] h=0 c#=113 <26f04cb5014ea50d>
   [O6] 0.194/345 [prior 0.944/180 amaf 0.111/2235 crit -0.314 vloss 0] h=0 c#=110 <26f05213014eaa58>
    [N8] 0.150/234 [prior 0.064/157 amaf 0.179/834 crit 0.000 vloss 0] h=0 c#=111 <26f052f8014eab3d>
    [O5] 0.188/85 [prior 0.079/126 amaf 0.163/498 crit -0.000 vloss 0] h=0 c#=111 <26f052eb014eab30>
   [M11] 0.007/144 [prior 0.942/173 amaf 0.060/1984 crit -0.403 vloss 0] h=0 c#=110 <26f0522e014eaa73>
    [L10] 0.007/135 [prior 0.080/125 amaf 0.007/411 crit -0.000 vloss 0] h=0 c#=111 <26f053e6014eac27>
   [F4] 0.018/114 [prior 0.923/130 amaf 0.070/1438 crit -0.000 vloss 0] h=0 c#=109 <26f05203014eaa48>
    [L10] 0.010/103 [prior 0.082/122 amaf 0.010/315 crit 0.000 vloss 0] h=0 c#=111 <26f054c5014ead04>
   [A3] 0.022/91 [prior 0.684/95 amaf 0.048/610 crit -0.978 vloss 0] h=0 c#=110 <26f051fe014eaa43>
    [B3] 0.025/81 [prior 0.082/122 amaf 0.024/246 crit -0.000 vloss 0] h=0 c#=111 <26f056cc014eaf08>
   [K16] 0.047/86 [prior 0.939/165 amaf 0.072/2183 crit -0.001 vloss 0] h=0 c#=109 <26f05242014eaa87>
    [P14] 0.039/76 [prior 0.077/130 amaf 0.043/235 crit 0.000 vloss 0] h=0 c#=111 <26f05c43014eb46e>
   [A1] 0.000/81 [prior 0.684/95 amaf 0.028/527 crit -1.000 vloss 0] h=0 c#=110 <26f051f0014eaa35>
    [B1] 0.000/70 [prior 0.078/128 amaf 0.000/213 crit 0.000 vloss 0] h=0 c#=111 <26f05730014eaf6a>
   [B1] 0.012/81 [prior 0.684/95 amaf 0.036/553 crit -0.988 vloss 0] h=0 c#=110 <26f051f1014eaa36>
    [A1] 0.015/66 [prior 0.071/140 amaf 0.014/207 crit 0.000 vloss 0] h=0 c#=111 <26f05aad014eb2dc>
   [pass] 0.000/75 [prior 0.500/20 amaf 0.000/0 crit -1.000 vloss 0] h=0 c#=111 <26f051ef014eaa34>
    [P14] 0.000/64 [prior 0.125/80 amaf 0.000/192 crit 0.000 vloss 0] h=0 c#=112 <26f058c9014eb0fe>
   [P14] 0.000/68 [prior 0.875/80 amaf 0.050/782 crit -0.456 vloss 0] h=0 c#=110 <26f0523c014eaa81>
    [O14] 0.000/57 [prior 0.069/145 amaf 0.000/177 crit 0.000 vloss 0] h=0 c#=113 <26f055b8014eadf6>
   [N14] 0.304/56 [prior 0.833/60 amaf 0.204/810 crit -0.643 vloss 0] h=0 c#=110 <26f0523a014eaa7f>
  [S4] 0.050/161 [prior 0.097/103 amaf 0.091/1688 crit -0.571 vloss 0] h=0 c#=113 <26f04cc2014ea51a>
   [O6] 0.050/60 [prior 0.949/195 amaf 0.035/254 crit -0.350 vloss 0] h=0 c#=110 <26f04d42014ea598>
 [N14] 0.941/882 [prior 0.944/178 amaf 0.730/3671 crit -0.042 vloss 0] h=0 c#=111 <26f04c86014ea4de>
  [N13] 0.950/358 [prior 0.051/198 amaf 0.956/1740 crit 0.000 vloss 0] h=0 c#=113 <26f05387014eabca>
   [F4] 0.970/237 [prior 0.923/130 amaf 0.969/797 crit -0.000 vloss 0] h=0 c#=108 <26f07286014eca71>
    [K1] 0.907/54 [prior 0.087/115 amaf 0.939/245 crit 0.101 vloss 0] h=0 c#=112 <26f07975014ed154>
   [O6] 0.911/90 [prior 0.949/195 amaf 0.938/518 crit -0.031 vloss 0] h=0 c#=109 <26f07296014eca81>
  [K15] 0.945/109 [prior 0.120/83 amaf 0.944/673 crit -0.014 vloss 0] h=0 c#=113 <26f0538d014eabd0>
   [O6] 0.927/55 [prior 0.944/180 amaf 0.930/200 crit -0.000 vloss 0] h=0 c#=109 <26f07459014ecc42>
  [J11] 0.978/89 [prior 0.303/99 amaf 0.977/393 crit 0.000 vloss 0] h=0 c#=114 <26f0537d014eabc0>
   [K16] 0.987/77 [prior 0.953/215 amaf 0.987/239 crit 0.000 vloss 0] h=0 c#=108 <26f073a4014ecb8d>
  [H10] 0.985/68 [prior 0.333/90 amaf 0.990/303 crit -0.000 vloss 0] h=0 c#=114 <26f05375014eabb8>
   [K16] 0.982/55 [prior 0.953/215 amaf 0.982/165 crit 0.000 vloss 0] h=0 c#=108 <26f074f8014ecce1>
  [K1] 0.906/64 [prior 0.133/75 amaf 0.931/788 crit 0.054 vloss 0] h=0 c#=113 <26f05346014eab89>
  [pass] 0.968/63 [prior 0.500/20 amaf 0.000/0 crit -0.032 vloss 0] h=0 c#=114 <26f05340014eab83>
 [A2] 0.010/191 [prior 0.960/248 amaf 0.341/1098 crit -0.859 vloss 0] h=0 c#=111 <26f04c45014ea49d>
  [A3] 0.011/182 [prior 0.050/200 amaf 0.011/552 crit -0.000 vloss 0] h=0 c#=112 <26f04fc9014ea819>
 [B13] 0.043/140 [prior 0.870/231 amaf 0.218/772 crit -0.829 vloss 0] h=0 c#=111 <26f04c83014ea4db>
  [B12] 0.042/118 [prior 0.035/285 amaf 0.039/383 crit 0.000 vloss 0] h=0 c#=113 <26f050e1014ea92d>
(avg score 1.495299/5010; dynkomi's 2.377203/644907 value 0.539632/645342)
*** WINNER is Q1 (16,1) with score 0.0958 (1461/5004:5004/5047 games), extra komi 0.000000
genmove in 2.45s (2061 games/s, 515 games/s/thread)
[5047] best 0.095825 xkomi 0.0 | seq  Q1  A2  O6  N8 | can b  Q1(0.096) N14(0.941)  A2(0.010) B13(0.043) 
(resign)
```

**Comment by lemonsqueeze** [comment]
I tried to recreate the same conditions, but so far i haven't been able to reproduce at all:

    cat gtp.txt | ./pachi -t =5000 threads=4,resign_threshold=0.25,dynkomi=linear:handicap_value=8%8:moves=150%150,max_tree_size=300,pondering=0

with this [gtp.txt](https://github.com/pasky/pachi/files/383413/gtp.txt) and same binary (pachi + dcnn, orig seki branch before PR cleanup).
This [moggy test](https://github.com/pasky/pachi/files/383453/moggy_test.txt) looks pretty consistent too.


Wondering what might cause something like this...
Doesn't look like the multithreaded bug, from what i've seen it either affects a given binary or not, so i'd be able to reproduce.
PRNG issue with fast_random() maybe ?

**Comment by lemonsqueeze** [comment]
I tried Bo Allen's [simple visual analysis](https://www.random.org/analysis) from random.org on fast_random().
Code is in my [random_test](https://github.com/lemonsqueeze/pachi/tree/random_test) branch (t-random directory). It uses sdl to generate a new random image for each key press:

![fast_random](https://cloud.githubusercontent.com/assets/2463904/17134019/c03bb31a-5328-11e6-95a8-32cd6354820b.png)

Not so random is it ?
It's also troubling that the same numbers get generated over and over again for each key press (only 64k of them it seems)

Out of curiosity i tried rand() which seems to be doing better:

![rand](https://cloud.githubusercontent.com/assets/2463904/17134026/c79ff9e0-5328-11e6-9928-ae80085cfda0.png)

**Comment by pasky** [comment]
Yes, I'm aware that fast_random() is pretty lousy, but since the random numbers also change meaning after every move, I was never too concerned about that.  I'd be convinced if you could show that Pachi misevaluates that situation wrong in N(=1000?) trials and not so with plain rand().  But intuitively I don't think the chances are that high.

It could be in theory also some issue with multi-threading, but I can't quite imagine an exact mechanism of that.

**Comment by lemonsqueeze** [comment]
That's what i was thinking too, it would take such an unlikely series of correlated random numbers to make something like that happen, seems impossible. But that would also explain why it's hard to reproduce ...

I'm thinking, we don't know the state of the generator when that genmove happened but we know the seed for this game. If the generator indeed loops over 64k numbers we could try them all, see what happens. It might trigger it reproduceably if it's indeed a random number issue. Multithreading makes it somewhat undeterministic though which is unfortunate ...

**Comment by lemonsqueeze** [comment]
Oh wow, this is a dynamic komi issue actually:
Extra komi was 3 during previous move, looks like 5000 sims is not enough to recover here (10000 sims would have been fine it seems). Is there something we can do to speed things up for low number of simulations ?

    $ while true ; do 
      ./pachi -t =5000 threads=4,max_tree_size=300,pondering=0,extra_komi=3 < gtp.txt 2>&1 | 
      grep WINNER ; done 
    *** WINNER is B13 (2,13) with score 0.0964 (1369/5003:5003/5116 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.0987 (1824/5004:5004/5108 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.0720 (1209/5004:5004/5103 games), extra komi 0.000000
    *** WINNER is A3 (1,3) with score 0.8053 (1171/5004:5004/5097 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.1025 (1336/5004:5004/5093 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.0725 (1311/5004:5004/5110 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.0781 (1447/5004:5004/5087 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.5226 (1923/5004:5004/5191 games), extra komi 0.000000
    *** WINNER is O4 (14,4) with score 0.5351 (1738/5003:5003/5081 games), extra komi 0.000000
    *** WINNER is A2 (1,2) with score 0.5813 (1323/5004:5004/5140 games), extra komi 0.000000
    *** WINNER is O4 (14,4) with score 0.8255 (974/5004:5004/5126 games), extra komi 0.000000
    *** WINNER is P5 (15,5) with score 0.0830 (1157/5004:5004/5112 games), extra komi 0.000000
    *** WINNER is L7 (11,7) with score 0.7554 (1014/5004:5004/5113 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.5003 (1861/5004:5004/5102 games), extra komi 0.000000
    *** WINNER is A12 (1,12) with score 0.0716 (1075/5004:5004/5119 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.4835 (1692/5004:5004/5120 games), extra komi 0.000000
    *** WINNER is B13 (2,13) with score 0.1395 (1276/5004:5004/5116 games), extra komi 0.000000
    *** WINNER is A12 (1,12) with score 0.6332 (1336/5004:5004/5136 games), extra komi 0.000000
    *** WINNER is A3 (1,3) with score 0.7640 (1021/5004:5004/5115 games), extra komi 0.000000

**Comment by lemonsqueeze** [comment]
Looks like dynkomi_interval is doing it:
Default value is 1000 and with 4 threads dynkomi_interval * threads = 4000 which is too much here.
Passing `dynkomi_interval=100` fixes it.

Maybe for fixed number of simulations we could initialize dykomi_interval based on number of simulations this way it'd always work ?


```C
void
uct_search_progress(struct uct *u, struct board *b, enum stone color,
		    struct tree *t, struct time_info *ti,
		    struct uct_search_state *s, int i)
{
	struct uct_thread_ctx *ctx = s->ctx;

	/* Adjust dynkomi? */
	int di = u->dynkomi_interval * u->threads;
	if (ctx->t->use_extra_komi && u->dynkomi->permove
	    && !u->pondering && di
	    && i > s->last_dynkomi + di) {
		s->last_dynkomi += di;
		floating_t old_dynkomi = ctx->t->extra_komi;
		ctx->t->extra_komi = u->dynkomi->permove(u->dynkomi, b, ctx->t);
		if (UDEBUGL(3) && old_dynkomi != ctx->t->extra_komi)
			fprintf(stderr, "dynkomi adjusted (%f -> %f)\n",
				old_dynkomi, ctx->t->extra_komi);
	}
```

**Comment by pasky** [comment]
That sounds like a great idea!

**Comment by lemonsqueeze** [comment]
Fixed by PR #73
